### DISCUSSION TITLE
Mobile-Project on .Net 8 fails to restore

### DISCUSSION THREAD

**Discussion by TopperDEL** [unoplatform/uno#14594] [Q&A] [answered]
I have a Mobile-project migrated to .Net8, but now get the following message in my Github Action:

![image](https://github.com/unoplatform/uno/assets/1833242/0a148d53-b142-4d7c-8573-b8304690ead1)

I was not yet able to build the project locally, as VS for Mac does not support .Net 8 and I was unable to setup VS Code correctly. So I've updated my build pipeline to use Net8 that was previously correctly building for .Net 7.

Any clues?

This is my csproj:
```
<Project Sdk="Microsoft.NET.Sdk">
	<PropertyGroup>
		<TargetFrameworks>net8.0-android;net8.0-ios;net8.0-maccatalyst</TargetFrameworks>
		<TargetFrameworks Condition="'$(OverrideTargetFrameworks)'!=''">$(OverrideTargetFrameworks)</TargetFrameworks>
		<SingleProject>true</SingleProject>
		<OutputType>Exe</OutputType>
		<!-- Display name -->
		<ApplicationTitle>MyApp</ApplicationTitle>
		<!-- App Identifier -->
		<ApplicationId>com.companyname.MyApp</ApplicationId>
		<ApplicationIdGuid>C207DEDB-...</ApplicationIdGuid>
		<!-- Versions -->
		<ApplicationDisplayVersion>1.0</ApplicationDisplayVersion>
		<ApplicationVersion>1</ApplicationVersion>

		<AndroidManifest>Android\AndroidManifest.xml</AndroidManifest>

		<!-- Debugger workaround https://github.com/dotnet/maui-samples/blob/8aa6b8780b12e97b157514c3bdc54bb4a13001cd/HelloMacCatalyst/HelloMacCatalyst.csproj#L7 -->
		<!-- <MtouchExtraArgs Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'maccatalyst'">$(MtouchExtraArgs) -setenv:MONO_THREADS_SUSPEND=preemptive</MtouchExtraArgs> -->
		<!-- Required for C# Hot Reload -->
		<UseInterpreter Condition="'$(Configuration)' == 'Debug' and $([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) != 'maccatalyst'">True</UseInterpreter>
		<IsUnoHead>true</IsUnoHead>
	</PropertyGroup>

	<ItemGroup>
		<PackageReference Include="Uno.Resizetizer" Version="1.2.1" />
		<PackageReference Include="Uno.WinUI" Version="5.0.48" />
		<PackageReference Include="CommunityToolkit.Mvvm" Version="8.2.2" />
		<PackageReference Include="Microsoft.Extensions.Logging.Console" Version="8.0.0" />
		<PackageReference Include="Uno.Extensions.Logging.OSLog" Version="1.7.0" />
		<PackageReference Include="Uno.UI.Adapter.Microsoft.Extensions.Logging" Version="5.0.48" />
		<PackageReference Include="Uno.WinUI.RemoteControl" Version="5.0.48" Condition="'$(Configuration)'=='Debug'" />
		<PackageReference Include="uplink.NET" Version="2.11.3183" />
		<PackageReference Include="uplink.NET.Mac" Version="2.11.3183" />
	</ItemGroup>
	<Choose>
		<When Condition="$(IsMacCatalyst)">
      <PropertyGroup>
        <!-- Configure the GC -->
        <MtouchExtraArgs>$(MtouchExtraArgs) --setenv=MONO_GC_PARAMS=soft-heap-limit=512m,nursery-size=64m,evacuation-threshold=66,major=marksweep,concurrent-sweep</MtouchExtraArgs>
        <!-- Required for unknown crash as of .NET 6 Mobile Preview 13 -->
        <MtouchExtraArgs>$(MtouchExtraArgs) --registrar:static</MtouchExtraArgs>
        <!-- Full globalization is required for Uno -->
        <InvariantGlobalization>false</InvariantGlobalization>
        <RuntimeIdentifier Condition="'$(RuntimeIdentifier)'==''">maccatalyst-x64</RuntimeIdentifier>
      </PropertyGroup>
      <PropertyGroup Condition="'$(Configuration)'=='Release'">
        <!-- https://github.com/xamarin/xamarin-macios/issues/14812 -->
        <MtouchExtraArgs>$(MtouchExtraArgs) --marshal-objectivec-exceptions:disable</MtouchExtraArgs>
      </PropertyGroup>
    </When>
	</Choose>
	<ItemGroup>
		<ProjectReference Include="..\MyApp\MyApp.csproj" />
	</ItemGroup>
	<ItemGroup>
	  <Folder Include="Resources\" />
	</ItemGroup>
	<ItemGroup>
	  <Page Include="..\MyApp\Resources\Styles.xaml" Link="Resources\Styles.xaml">
	    <Generator>MSBuild:Compile</Generator>
	  </Page>
	</ItemGroup>
	<Import Project="..\MyApp.Base\base.props" />
</Project>
```

**Comment by jeromelaban** [comment]
Thanks for the question. This generally means that workloads have not been installed. Did you run uno-check in your build?

**Reply by TopperDEL** [reply]
I have this step before in my pipeline:

`dotnet workload install maui --ignore-failed-sources`

**Reply by jeromelaban** [reply]
It's generally best to use uno-check, as it will pin a specific version to your build (and avoid later changes in hosted agents underlying versions).

You can also [troubleshoot your build with binlogs](https://platform.uno/docs/articles/uno-builds-troubleshooting.html).

**Reply by TopperDEL** [reply]
I just added a step to install and call uno-check. It does indeed install plenty of workloads, but the error afterwars is still the same. I'll give the binlog a try - but if you have any other idea let me know.

**Reply by TopperDEL** [reply]
I could create a binlog and I even can reproduce the output with a fresh Maui-Project.
[MauMacos.zip](https://github.com/unoplatform/uno/files/13525887/MauMacos.zip)
[binlog.zip](https://github.com/unoplatform/uno/files/13525891/binlog.zip)

If I create a readable ouput from the binlog using `msbuild /bl:msbuild.binlog`I get the error NETSDK1013: The TargetFramework-Value "net8.0-maccatalyst" was not found/recognized (it is in german on my machine).

Any more ideas? Workloads seem to be ok.

**Reply by limeniye** [reply]
I suggest you to update and run ```maui-check```

**Reply by TopperDEL** [reply]
maui-check instead of uno-check?

**Reply by TopperDEL** [reply]
After reinstalling xcode on my mac, the above "MauMacos" (typo)-project is building. But my main project isn't. Neither on the mac, nor in my github-action-pipeline. Always the same error.

`/Users/runner/.dotnet/sdk/8.0.100/NuGet.targets(156,5): error : Invalid framework identifier ''.`

**Reply by TopperDEL** [reply]
This is my updated csproj:

```
<Project Sdk="Microsoft.NET.Sdk">
	<PropertyGroup>
		<TargetFrameworks>net8.0-maccatalyst</TargetFrameworks>
		<TargetFrameworks Condition="'$(OverrideTargetFrameworks)'!=''">$(OverrideTargetFrameworks)</TargetFrameworks>
		<SingleProject>true</SingleProject>
		<OutputType>Exe</OutputType>
		<!-- Display name -->
		<ApplicationTitle>MyApp</ApplicationTitle>
		<!-- App Identifier -->
		<ApplicationId>com.MyApp</ApplicationId>
		<ApplicationIdGuid>MYGUID</ApplicationIdGuid>
		<!-- Versions -->
		<ApplicationDisplayVersion>1.0</ApplicationDisplayVersion>
		<ApplicationVersion>1</ApplicationVersion>

		<!-- Debugger workaround https://github.com/dotnet/maui-samples/blob/8aa6b8780b12e97b157514c3bdc54bb4a13001cd/HelloMacCatalyst/HelloMacCatalyst.csproj#L7 -->
		<!-- <MtouchExtraArgs Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'maccatalyst'">$(MtouchExtraArgs) -setenv:MONO_THREADS_SUSPEND=preemptive</MtouchExtraArgs> -->
		<!-- Required for C# Hot Reload -->
		<UseInterpreter Condition="'$(Configuration)' == 'Debug' and $([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) != 'maccatalyst'">True</UseInterpreter>
		<IsUnoHead>true</IsUnoHead>
	</PropertyGroup>

	<ItemGroup>
		<PackageReference Include="Uno.Resizetizer" Version="1.2.1" />
		<PackageReference Include="Uno.WinUI" Version="5.0.48" />
		<PackageReference Include="CommunityToolkit.Mvvm" Version="8.2.2" />
		<PackageReference Include="Microsoft.Extensions.Logging.Console" Version="8.0.0" />
		<PackageReference Include="Uno.Extensions.Logging.OSLog" Version="1.7.0" />
		<PackageReference Include="Uno.UI.Adapter.Microsoft.Extensions.Logging" Version="5.0.48" />
		<PackageReference Include="Uno.WinUI.RemoteControl" Version="5.0.48" Condition="'$(Configuration)'=='Debug'" />
		<PackageReference Include="uplink.NET" Version="2.11.3183" />
		<PackageReference Include="uplink.NET.Mac" Version="2.11.3183" />
	</ItemGroup>
	<Choose>
		<When Condition="$(IsMacCatalyst)">
      <PropertyGroup>
        <!-- Configure the GC -->
        <MtouchExtraArgs>$(MtouchExtraArgs) --setenv=MONO_GC_PARAMS=soft-heap-limit=512m,nursery-size=64m,evacuation-threshold=66,major=marksweep,concurrent-sweep</MtouchExtraArgs>
        <!-- Required for unknown crash as of .NET 6 Mobile Preview 13 -->
        <MtouchExtraArgs>$(MtouchExtraArgs) --registrar:static</MtouchExtraArgs>
        <!-- Full globalization is required for Uno -->
        <InvariantGlobalization>false</InvariantGlobalization>
        <RuntimeIdentifier Condition="'$(RuntimeIdentifier)'==''">maccatalyst-x64</RuntimeIdentifier>
      </PropertyGroup>
      <PropertyGroup Condition="'$(Configuration)'=='Release'">
        <!-- https://github.com/xamarin/xamarin-macios/issues/14812 -->
        <MtouchExtraArgs>$(MtouchExtraArgs) --marshal-objectivec-exceptions:disable</MtouchExtraArgs>
      </PropertyGroup>
    </When>
	</Choose>
	<ItemGroup>
		<ProjectReference Include="..\MyApp\MyApp.csproj" />
	</ItemGroup>
	<ItemGroup>
	  <Folder Include="Resources\" />
	</ItemGroup>
	<ItemGroup>
	  <Page Include="..\MyApp\Resources\Styles.xaml" Link="Resources\Styles.xaml">
	    <Generator>MSBuild:Compile</Generator>
	  </Page>
	</ItemGroup>
	<Import Project="..\MyApp.Base\base.props" />
</Project>
```

**Reply by TopperDEL** [reply]
I am completely lost here. I've updated the project file to the latest version I get from the templates. Also `Directory.Build.props` and `target` - nothing changes.
Calling ´dotnet release` leads to Ìnvalid Target framework`. Calling `msbuild` says `NETSDK1013: The TargetFramework-value "net8.0-maccatalyst" was not recognized'"`.

I don't know what I could further check here.

**Reply by TopperDEL** [reply, marked as answer]
I think I found my issue. In my "shared"-project I was missing the TargetFrameworks in case I'm not building for windows. I removed the other targets anyhow after starting from the template and forgot them. Now I wanted to buidl for the mac and got the "TargetFramework-empty"-error - but unfortunately on the Head-project, not on the shared one. So it lead me to the wrong project.

**Comment by spouliot** [comment]
> I was unable to setup VS Code correctly

@TopperDEL can you open an issue or another discussion to let us know about that problem ? thanks!

**Reply by TopperDEL** [reply]
I'll give it a try again. But maybe I'm to lost with VS code itself. It....feels strange, anyhow.

**Reply by TopperDEL** [reply]
Ok, just did a new try. So calling "code" on the Terminal does not work. But I can open the folder by myself, this is just a pathing issue.

Then I'm struggling with starting the Mobile project. [here](https://platform.uno/docs/articles/create-an-app-vscode.html?tabs=catalystdebug) it states "In the status bar select the MyApp.Mobile project". But I do not have that in the status bar. I cannot switch it:
![image](https://github.com/unoplatform/uno/assets/1833242/d1441c7e-9163-4468-b7e2-9f9f9bbfb5de)

Hitting F5 or selecting "Start debugging" from the Menu does not lead to anything. I don't see any error or any build process starting up.

Once again: I'm a complete noob regarding VScode. Maybe I'm too dumb or too old.... :)

**Reply by spouliot** [reply]
> So calling "code" on the Terminal does not work. 

This needs to be setup inside VS Code. See https://platform.uno/docs/articles/get-started-vscode.html

> 3. Configure VS Code to start from the command-line using the code command. This can be configured by following [these instructions](https://code.visualstudio.com/docs/editor/command-line#_launching-from-command-line).



> But I do not have that in the status bar. 

In your screenshot you can see the `Uno Platform` logs mention:

> [Error] To enable most features, 'useOmnisharp' needs to be enabled in the settings. See https://aka.platform.uno/vscode-omnisharp-devkit for more details.

This is because the new C# extension (Microsoft) does not use OmniSharp by default anymore. However MS does not yet offer the API we need when C# DevKit (another MS extension) is used. So you need to manually configure the C# extension to use OmniSharp.

Hopefully we'll soon have an update to allow the use of C# DevKit with the Uno extension.

**Reply by TopperDEL** [reply]
Thanks, that helped alot. At least I get a new uno-app to open and build. But I still don#t get warm with vs code.

- I don't like the way files are presented. I cannot quickly recognize what a text in the structure is - is it a cs-file, is it XAML? I assume I can change this by editing the settings or by installing an extension. But it makes it hard to switch, at least for me.
- How do I add nuget packages?
- Building a project takes quite some time - but I guess I would need to add kind of a "solution filter" or what, so that only that gets built what I need currently.

I don't want to bash VScode, though I really have huge troubles with it. I only need it currently as I need to fix a bug on Mac and my net8-project is not supported from VS for mac anymore as it seems.

Thanks for your help, though!

**Reply by jeromelaban** [reply]
Note that this experience will change somehow with C# dev kit, where an actual solution explorer is available in our next vscode update with C# dev kit compatibility.

**Reply by spouliot** [reply]
> I don't like the way files are presented.

As @jeromelaban mentioned this will get better with C# DevKit as it provides a Solution Explorer that is similar to the one found in VS (Win or Mac).

> How do I add nuget packages?

I think it's in the plan for future versions C# DevKit. Right now you can an extension like https://marketplace.visualstudio.com/items?itemName=patcx.vscode-nuget-gallery to manage NuGet packages

> Building a project takes quite some time 

It should not take longer. You can select (in the status bar) the active project and only related projects should be built (it's all done by `dotnet`, not the extension).

>  is not supported from VS for mac anymore as it seems.

Yes, a lot of VSMac users will need migrate to other tools (or platform) to continue their work :(
